### DISCUSSION TITLE
onedrive not synching after ubuntu upgrade

### DISCUSSION THREAD

**Discussion by tbgoncalves** [abraunegg/onedrive#2129] [Q&A] [answered]
Hello. Three days ago I upgraded from Ubuntu 20.04 to 22.04, and, since then, onedrive is not synchronizing. I had it set up such that it was synchronizing all my files in my OneDrive into a local folder on my laptop - it worked greatly both ways! Now it seems it might still be able to download, but it is not uploading. It runs into some "No such file or directory" problem; this is what I get when I try ```onedrive --synchronize```:

```
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Downloading file Documentos/Doutoramento/Literatura/LiteratureRecordSynthesis.tex ... done.
Retry sync count: 1: Documentos/Doutoramento/UniversoPrimitivo-TópicosAvançados/ExcerciseSheet3code_TGoncalves/venv/bin/python3.8: No such file or directory
Syncing changes from OneDrive ...
Retry sync count: 2: Documentos/Doutoramento/UniversoPrimitivo-TópicosAvançados/ExcerciseSheet3code_TGoncalves/venv/bin/python3.8: No such file or directory
Syncing changes from OneDrive ...
Giving up on sync after three attempts: Documentos/Doutoramento/UniversoPrimitivo-TópicosAvançados/ExcerciseSheet3code_TGoncalves/venv/bin/python3.8: No such file or directory
std.file.FileException@/usr/lib/ldc/x86_64-linux-gnu/include/d/std/file.d(2055): Documentos/Doutoramento/UniversoPrimitivo-TópicosAvançados/ExcerciseSheet3code_TGoncalves/venv/bin/python3.8: No such file or directory
----------------
??:? @trusted bool std.file.cenforce!(bool).cenforce(bool, scope const(char)[], scope const(char)*, immutable(char)[], ulong) [0x55f033b165a6]
??:? @safe uint std.file.getAttributes!(immutable(char)[]).getAttributes(immutable(char)[]) [0x55f033b16622]
??:? [0x55f033a89bbd]
??:? [0x55f033a8a761]
??:? [0x55f033a822ca]
??:? [0x55f033a78ec2]
??:? [0x55f033a736c6]
??:? [0x55f033ac52e0]
??:? [0x55f033abb666]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7f89aa22ea4b]
??:? _d_run_main2 [0x7f89aa22e867]
??:? _d_run_main [0x7f89aa22e6bd]
??:? [0x7f89a9e4dd8f]
??:? __libc_start_main [0x7f89a9e4de3f]
??:? [0x55f033a4c104]
std.net.curl.CurlException@std/net/curl.d(4402): A libcurl function was given a bad argument on handle null
----------------
??:? pure @safe noreturn std.exception.bailOut!(std.net.curl.CurlException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x7f89aa5adad7]
??:? void std.net.curl.Curl.clear(etc.c.curl.CurlOption) [0x7f89aa50e363]
??:? [0x55f033ab7ea5]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7f89aa22ea4b]
??:? _d_run_main2 [0x7f89aa22e867]
??:? _d_run_main [0x7f89aa22e6bd]
??:? [0x7f89a9e4dd8f]
??:? __libc_start_main [0x7f89a9e4de3f]
??:? [0x55f033a4c104]
```

I've tried to uninstall onedrive and reinstalling it following the instructions on https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md, using OpenSuSE Build Service repository. But, it is still not working.

In fact, initially, I had some trouble uninstalling onedrive - I guess it was because over a year ago when I first installed it, I might have not done it using ```sudo apt-get install```. Eventually, I think I was able to, using ```make uninstall```; and because I am not yet very proficient, I also tried by manually deleting the onedrive folders (such as the one inside .config) - I'm not sure if that's something I could do. One thing or the other seemed to do the trick. 

Then, I did a clean installation, but I am running into the same problem as detailed above. I also tried to delete the directory in which it is finding a problem, both online and locally, but in a following try to synchronize it just breaks in a different directory. 

This is the onedrive version I now have:
```onedrive v2.4.20-1+np1```
The command ```onedrive --display-config``` returns: 
```
onedrive version                             = v2.4.20-1+np1
Config path                                  = /home/tibago/.config/onedrive
Config file found in config path             = false
Config option 'sync_dir'                     = /home/tibago/OneDrive
Config option 'enable_logging'               = false
Config option 'log_dir'                      = /var/log/onedrive/
Config option 'disable_notifications'        = false
Config option 'min_notify_changes'           = 5
Config option 'skip_dir'                     = 
Config option 'skip_dir_strict_match'        = false
Config option 'skip_file'                    = ~*|.~*|*.tmp
Config option 'skip_dotfiles'                = false
Config option 'skip_symlinks'                = false
Config option 'monitor_interval'             = 300
Config option 'monitor_log_frequency'        = 5
Config option 'monitor_fullscan_frequency'   = 12
Config option 'dry_run'                      = false
Config option 'upload_only'                  = false
Config option 'download_only'                = false
Config option 'local_first'                  = false
Config option 'check_nosync'                 = false
Config option 'check_nomount'                = false
Config option 'resync'                       = false
Config option 'resync_auth'                  = false
Config option 'classify_as_big_delete'       = 1000
Config option 'disable_upload_validation'    = false
Config option 'bypass_data_preservation'     = false
Config option 'no_remote_delete'             = false
Config option 'remove_source_files'          = false
Config option 'sync_dir_permissions'         = 700
Config option 'sync_file_permissions'        = 600
Config option 'space_reservation'            = 52428800
Config option 'application_id'               = 
Config option 'azure_ad_endpoint'            = 
Config option 'azure_tenant_id'              = common
Config option 'user_agent'                   = 
Config option 'force_http_11'                = false
Config option 'debug_https'                  = false
Config option 'rate_limit'                   = 0
Config option 'operation_timeout'            = 3600
Config option 'sync_root_files'              = false
Selective sync 'sync_list' configured        = false
Config option 'sync_business_shared_folders' = false
Business Shared Folders configured           = false
Config option 'webhook_enabled'              = false
```
And ```onedrive --display-sync-status``` returns:
```
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Local directory is out of sync with OneDrive
Approximate data to download from OneDrive: 1923 KB
```
My system details are:
```
Linux <pc_name> 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.1 LTS
Release:	22.04
Codename:	jammy
```
```
curl --version
curl 7.84.0 (x86_64-pc-linux-gnu) libcurl/7.84.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 zstd/1.4.4 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3 libgsasl/1.8.1 OpenLDAP/2.4.49
Release-Date: 2022-06-27
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli gsasl GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```

What could I do to make it work again? Please, let me know if more details are needed, and if there is something I'm not doing properly.

**Comment by abraunegg** [comment, marked as answer]
@tbgoncalves 
> What could I do to make it work again?

Potentially, it 'sounds' like you *still* have an old binary lurking around - as, the error message and crash you are getting sounds a hell of a lot like this issue: https://github.com/abraunegg/onedrive/issues/1915 - and this was resolved in v2.4.17+

Suggestion to resolve:
1. Uninstall the application using `sudo apt remove onedrive`
2. Reboot
3. Search your entire system for a 'onedrive' binary - using `find` or similar .. or even `which onedrive` to ensure that there is no old version lying around. Keep searching your platform if you find one, and remove it manually.
4. Once no further binary can be found, perform a further reboot to ensure that there is no copy running in memory
5. Reinstall the application following https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md

If you then are *still* having an issue - you have found a *new* bug. Please follow the correct process to raise a bug report.

**Reply by tbgoncalves** [reply]
@abraunegg, many thanks for your prompt reply. 

I have done something that broke my system that made the computer inoperable. So, unfortunately, it will be some time before I can try what you suggest. Once I do - if I can get the system back to working order - I will let you know if it worked for the onedrive. 

Thank you.

**Reply by abraunegg** [reply]
@tbgoncalves 
No problem. It potentially sounds like a re-install of your Operating System .. and maybe an opportunity to use something better than Ubuntu.

**Reply by tbgoncalves** [reply]
@abraunegg, do you have an OS recommendation to replace Ubuntu? One not overcomplicated as I'm not yet completely familiar with Linux (clearly, since I managed to break it 😅) Thanks

**Reply by abraunegg** [reply]
@tbgoncalves 
A Linux distro discussion can cause all sorts of discussions (heated or otherwise). I have been using Linux since 'boot' and 'root' floppy disk days ......

If you are going to look at replacing Ubuntu, I would investigate:
* Fedora 
* Rocky Linux
* RedHat Enterprise (given the change in licensing allowing personal use up to 8 devices, fully supported)
* Manjaro (which is based on ArchLinux)
* ArchLinux
* Gentoo
* Slackware

Personally I would steer clear of anything Ubuntu based, or derived from Ubuntu ... but thats my choice.

**Reply by tbgoncalves** [reply]
@abraunegg
I have used Timeshift with a live Ubuntu USB to restore my system to how it was on Monday, when it was still all working (uff, what a relief!). So now I have it as it was before updating Ubuntu: I am running ```Ubuntu 20.04.5 LTS``` with ```onedrive v2.4.12-9-g6201938```, and onedrive seems to be working normally. 

Shall I, then, out of curiosity, do the same as before, i.e. updating to Ubuntu 22.04, see if onedrive gets the same errors, and check if the solution you suggested fixes it? 

Or shall I simply update the onedrive version, without making changes to the system? 
What's the proper way to update from this version of onedrive? Do I need to uninstall the old one first? How do I do it? Since when I run ```$ sudo apt remove onedrive``` what I get is ```Package 'onedrive' is not installed, so not removed```.

Thank you.

**Comment by abraunegg** [comment]
@tbgoncalves 
> `onedrive v2.4.12-9-g6201938` 

This means you manually installed ... so you need to manually uninstall. You are running a 100% unsuppprted client. Please follow the directions to manually uninstall the client and make sure it is properly removed before proceeding any further.

Once removed, follow the *correct* instructions to install on Ubuntu to update your client correctly for Ubuntu 20.x

Once done, then do your upgrade to Ubuntu 22.x and then upgrade your onedrive client following the correct Ubuntu install instructions.

**Reply by tbgoncalves** [reply]
@abraunegg 
It doesn't seem to be working: 

> you need to manually uninstall

Check

> follow the correct instructions to install on Ubuntu to update your client correctly for Ubuntu 20.x

Check. The updated onedrive was working on Ubuntu 20.x

> then do your upgrade to Ubuntu 22.x

check

>and then upgrade your onedrive client following the correct Ubuntu install instructions

Tried to follow https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md: I've removed the previous version and repository, updated the system... etc... and then added repository for Ubuntu 22.x and installed onedrive  
When I run ```onedrive --synchronize``` it crashed as it was doing initially when it prompted my first question here above.

I've tried to uninstall it once again; deleted files named with onedrive; After that I had some problems when trying to install the client again:

```
$ sudo apt install onedrive
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies.
 onedrive : Depends: libphobos2-ldc-shared94 (>= 1:1.24.0) but it is not installable
E: Unable to correct problems, you have held broken packages.

$ sudo apt install libphobos2-ldc-shared94
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package libphobos2-ldc-shared94 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'libphobos2-ldc-shared94' has no installation candidate
```
I tried installing ```libphobos2-ldc-shared-dev``` and do another system update and reboot. Then it installed the one drive client without a problem.

Still, at the moment, it is crashing in a similar way as initially.

**Reply by abraunegg** [reply]
@tbgoncalves 
> Still, at the moment, it is crashing in a similar way as initially.

If you have removed the application .. there is *no way* it can be crashing ... as .. it should not exist ... so you are not removing it correctly.

Please do the following:
```
cd ~
sudo apt install build-essential
sudo apt install libcurl4-openssl-dev libsqlite3-dev pkg-config git curl ldc
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
./configure
sudo make uninstall
```

This should force the application to correctly uninstall your manual install on your Ubuntu system:
```
rm -f /usr/local/bin/onedrive
rm -f /usr/local/share/man/man1/onedrive.1
rm -f /usr/local/etc/logrotate.d/onedrive
rm -f /lib/systemd/system/onedrive*.service
rm -f /usr/lib/systemd/user/onedrive*.service
for i in README.md config LICENSE CHANGELOG.md docs/Docker.md docs/INSTALL.md docs/SharePoint-Shared-Libraries.md docs/USAGE.md docs/BusinessSharedFolders.md docs/advanced-usage.md docs/application-security.md ; do rm -f /usr/local/share/doc/onedrive/$i ; done
```

Then you need to *remove* the GitHib clone directory and these packages as well ...............
```
cd ~
rm -rf ./onedrive
sudo apt remove libcurl4-openssl-dev libsqlite3-dev pkg-config git curl ldc
sudo apt remove build-essential
sudo apt-get autoremove -y
sudo apt-get autoclean -y
```

For reference - on Ubuntu 20.x and 22.x a manual install will put the application binary in: `/usr/local/bin/onedrive`

Make sure this 'binary' does *not* exist on your system.

Additionally, because you are unsure what you are doing ... where is the 'default' result for 'onedrive' coming from?
```
which onedrive
```

If the client does not exist on your system, this command will not find it.


> I tried installing libphobos2-ldc-shared-dev and do another system update and reboot.

When installing the client *correctly* following the explicit directions as per https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md for Ubuntu 22.x - all package dependancies are correctly met at package installation time.

Again - you are potentially *not* installing the application correctly ..... please follow the directions carefully and accurately. 

Below is a correct installation sequence on Ubuntu 22.x following the documentation:
```
Welcome to Ubuntu 22.04.1 LTS (GNU/Linux 5.15.0-47-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

0 updates can be applied immediately.

Last login: Sun Sep 18 10:30:03 2022 from X.X.X.X
alex@ubuntu-22-04:~$ wget -qO - https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04/Release.key | gpg --dearmor | sudo tee /usr/share/keyrings/obs-onedrive.gpg > /dev/null
[sudo] password for alex: 
alex@ubuntu-22-04:~$ echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/obs-onedrive.gpg] https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04/ ./" | sudo tee /etc/apt/sources.list.d/onedrive.list
deb [arch=amd64 signed-by=/usr/share/keyrings/obs-onedrive.gpg] https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04/ ./
alex@ubuntu-22-04:~$ sudo apt-get update
Hit:1 http://au.archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://au.archive.ubuntu.com/ubuntu jammy-updates InRelease [114 kB]
Get:3 http://au.archive.ubuntu.com/ubuntu jammy-backports InRelease [99.8 kB]
Get:4 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]      
Get:5 http://au.archive.ubuntu.com/ubuntu jammy-updates/main amd64 DEP-11 Metadata [93.0 kB]
Get:6 http://au.archive.ubuntu.com/ubuntu jammy-updates/universe amd64 DEP-11 Metadata [240 kB]               
Get:7 http://au.archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 DEP-11 Metadata [940 B]                           
Get:8 https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04 ./ InRelease [1,604 B]
Get:9 http://au.archive.ubuntu.com/ubuntu jammy-backports/universe amd64 DEP-11 Metadata [12.3 kB]
Get:10 http://security.ubuntu.com/ubuntu jammy-security/main amd64 DEP-11 Metadata [13.1 kB]
Get:11 https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04 ./ Packages [807 B]
Get:12 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 DEP-11 Metadata [12.1 kB]
Fetched 698 kB in 3s (251 kB/s)              
Reading package lists... Done
alex@ubuntu-22-04:~$ sudo apt install onedrive
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libllvm11 libphobos2-ldc-shared98
The following NEW packages will be installed:
  libllvm11 libphobos2-ldc-shared98 onedrive
0 to upgrade, 3 to newly install, 0 to remove and 0 not to upgrade.
Need to get 21.8 MB of archives.
After this operation, 92.7 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://au.archive.ubuntu.com/ubuntu jammy/universe amd64 libllvm11 amd64 1:11.1.0-6 [19.6 MB]
Get:3 http://au.archive.ubuntu.com/ubuntu jammy/universe amd64 libphobos2-ldc-shared98 amd64 1:1.28.0-1ubuntu1 [1,487 kB]
Get:2 https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04 ./ onedrive 2.4.20-1+np1 [722 kB]
Fetched 21.8 MB in 4s (4,854 kB/s)
Selecting previously unselected package libllvm11:amd64.
(Reading database ... 176774 files and directories currently installed.)
Preparing to unpack .../libllvm11_1%3a11.1.0-6_amd64.deb ...
Unpacking libllvm11:amd64 (1:11.1.0-6) ...
Selecting previously unselected package libphobos2-ldc-shared98:amd64.
Preparing to unpack .../libphobos2-ldc-shared98_1%3a1.28.0-1ubuntu1_amd64.deb ...
Unpacking libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...
Selecting previously unselected package onedrive.
Preparing to unpack .../onedrive_2.4.20-1+np1_amd64.deb ...
Unpacking onedrive (2.4.20-1+np1) ...
Setting up libllvm11:amd64 (1:11.1.0-6) ...
Setting up libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...
Setting up onedrive (2.4.20-1+np1) ...
Processing triggers for man-db (2.10.2-1) ...
Processing triggers for libc-bin (2.35-0ubuntu3.1) ...
alex@ubuntu-22-04:~$ 


```

If you are not getting this sort of output - you are doing something wrong.

**Reply by tbgoncalves** [reply]
@abraunegg 
It seems to be working. Thank you very much for your patience and explanations.
